modeshape-jcr/src/main/java/org/modeshape/jcr/txn/SynchronizedTransactions.java

```diff
@@ -66,35 +68,49 @@ public SynchronizedTransactions( TransactionManager txnMgr, Cache localCache ) {
 
     @Override
     public Transaction currentTransaction() {
-        return ACTIVE_TRANSACTION.get();
+        Transaction localTx = LOCAL_TRANSACTION.get();
+        if (localTx != null) {
+            return localTx;
+        }
+        try {
+            javax.transaction.Transaction txn = txnMgr.getTransaction();
+            if (txn == null) {
+                // no active transaction
+                return null;
+            }
+            GlobalTransaction ispnTx = transactionTable.getGlobalTransaction(txn);
+            return ispnTx != null ? activeTransactionsByISPNGlobalTxId.get(ispnTx.getId()) : null;
+        } catch (SystemException e) {
+            logger.debug(e, "Cannot determine if there is an active transaction or not");
+            return null;
+        }
     }
 
     @Override
     public Transaction begin() throws NotSupportedException, SystemException {
         // check if there isn't an active transaction already
-        Transaction result = ACTIVE_TRANSACTION.get();
+        Transaction result = LOCAL_TRANSACTION.get();
         if (result != null) {
-            // we have an existing transaction so depending on the type we either need to be aware of nesting
-            // or return as-is
+            // we have an existing local transaction so we need to be aware of nesting
             if (logger.isTraceEnabled()) {
                 logger.trace("Found active ModeShape transaction '{0}' ", result);
             }
-            return result instanceof NestableThreadLocalTransaction ?
-                   ((NestableThreadLocalTransaction) result).begin() :
-                   result;
+            return ((NestableThreadLocalTransaction) result).begin();
```

Comment:
>NestedThreadLocalTransaction ntlt = LOCAL_TRANSACTION.get();
if ( ntlt != null ) {
    ...
    return ntlt.begin();
}
Transaction result = ntlt;

I understand what you're saying, but [this](https://github.com/ModeShape/modeshape/blob/master/modeshape-jcr/src/main/java/org/modeshape/jcr/txn/SynchronizedTransactions.java#L133) part of the code needs to always be executed because it adds extra (non trivial) logic for debugging purposes. Because of this, the method can't "short-circuit" and exit and has to make sure that part of the code is always executed, regardless of the transaction type.